### DISCUSSION TITLE
New IPA phonemizer engine - Goruut.

### DISCUSSION THREAD

**Discussion by neurlang** [coqui-ai/TTS#3794] [Show and tell]
Hello I've made a new IPA phonemizer engine, in golang. It is available online at [hashtron.cloud](https://hashtron.cloud) or self-hosted under a MIT license. [github](https://github.com/neurlang/goruut)

# What works:

## Supported Languages

*    English
*    Spanish
*    German
*    Arabic
*    Czech
*    Slovak

**Looking for more languages to add!**

## Target languages

* IPA
* Espeak
* Antvaset.com

# What doesn't  work:

* Currently dates, times, or anything with numbers doesn't work (fix planned)
* Slashes, dashes etc or words separated by anything other than spaces.
* More punctuation needs to be trimmed.

Feedback welcomed.

**Comment by neurlang** [comment]
# goruut phonemizer

## Supported languages

English, Spanish, Portuguese, German, French, Italian, Arabic, Farsi, Luxembourgish, Dutch, Russian, Swedish, Czech, Slovak

## Planned languages

Swahili, Japanese

**Looking for more languages to add!**

## Try it online

[https://www.hashtron.cloud/](https://www.hashtron.cloud/)

**Comment by komainuone** [comment]
Hello,
how does the new IPA phonemizer engine work? Is this documented anywhere? Can we add new languages easily?
Thank you.

**Reply by neurlang** [reply]
First of all, thank you for the good questions.

The IPA phonemizer searches the word's longest prefix match in the [language's grammar](https://github.com/neurlang/goruut/blob/master/dicts/spanish/language.json). Once a match is identified, the correct pronunciation is selected by a [hashtron machine learning model](https://github.com/neurlang/classifier). The process is then repeated for the remaining phones in the word. The performance is on the order of milliseconds for the whole sentence.

Adding a new language has been documented, the dict folder now does have a [readme of how to add a new language to goruut](https://github.com/neurlang/goruut/blob/master/dicts/README.md). Sharing a dataset with the community is fully **optional**.

**Comment by martinarisk** [comment]
I would like to upvote this. Using goruut 0.4.0 and the quality is very good for both phonemizer and reverse phonemizer.

**Reply by eginhard** [reply]
You can use the code in https://github.com/idiap/coqui-ai-TTS/pull/88 to use goruut with Coqui.

**Comment by neurlang** [comment]
# Goruut/Pygoruut v0.6.1

* 139 languages support
* fix compilation on 386
* retrained all models for forward / reverse phonemization (breaking change, new minor version)
* new transformer based models for homograph word inference (english)
* Handle numerics in Arabic and English
* working study initiation from null, {} map in language.json
* Add groups of same diacritics to hebrew2
* analysis2 normalization/diacritics sorting
* Fixed -overwrite in homograph scripts
* Split on hanzi/nonhanzi boundary
* don't crash if language dict not found in phonemization steps
* User interface
  * toucan TTS integration
  * word editor
* New Languages:
  * cantonese
  * minnan/taiwanese
  * minnan/hokkien
